### ISSUE TITLE
Use a yield iterator for ResolveFieldContext.Path

### ISSUE THREAD

**Issue by johnrutherford** [graphql-dotnet/graphql-dotnet#1147]
Currently during the document execution, string arrays are created to set the `Path` property on ResolveFieldContexts. This results in a lot of arrays being allocated for a property that is probably very rarely used. 

Since the property is `IEnumerable<string>`, we should use a yield iterator to generate the enumerable when it's actually needed. This should be pretty simple to do by walking up the execution node ancestors and then doing a `Reverse()`.

**Comment by Shane32** [comment]
We could also just change AppendPath as follows:
``` C#
        public static IEnumerable<string> AppendPath(IEnumerable<string> path, string pathSegment)
        {
            return path.Append(pathSegment);
        }
```
The Enumerable.Append code does not actually clone the array, but rather as it is enumerated, the proper values are returned.

**Comment by sungam3r** [comment]
But Enumerable.Append yet creates new object - iterator instance:
```c#
 public static IEnumerable<TSource> Append<TSource>(this IEnumerable<TSource> source, TSource element)
    {
      if (source == null)
        throw Error.ArgumentNull(nameof (source));
      Enumerable.AppendPrependIterator<TSource> appendPrependIterator = source as Enumerable.AppendPrependIterator<TSource>;
      if (appendPrependIterator != null)
        return (IEnumerable<TSource>) appendPrependIterator.Append(element);
      return (IEnumerable<TSource>) new Enumerable.AppendPrepend1Iterator<TSource>(source, element, true);
    }
```

**Comment by Shane32** [comment]
True, it's not lazy loaded per the suggestion, which would probably be best.  I just thought this solution was better than what's there now, and very simple to implement.

**Comment by sungam3r** [comment]
Honestly, I don’t think it will be easy. Intuition suggests that in order to optimize the memory allocation we will have to rethink a much larger amount of code, not only `Path`. This object is passed on to other objects, and they are passed on to others.

**Comment by Earthmark** [comment]
I believe list in the core framework gets around this by returning a struct instead of an interface, as long as you don't cast it to IEnumerable it's free. Doing foreach on it ends up using the struct directly, all it cares about is that the iterator calls line up.

**Comment by sungam3r** [comment]
You describe a known behavior, but it is not related to this issue. In any case, I just wanted to say that we need to look at the problem more broadly, without wasting energy on local optimizations. Algorithms should initially be designed taking into account the peculiarities of memory allocations, otherwise it will then be difficult to optimize them without violating an already created design that has been working for years.

**Comment by sungam3r** [comment]
Fixed in #1533
